### PR TITLE
perf: avoid memory leak issue #1834

### PR DESCRIPTION
**Pull request by tonyyip1969**
Avoid memory leak by remove the exclusive parameter and logic in the code block on AddEventListener method in DomEventService class.

The following are the components affected:
components/affix/Affix.razor.cs
components/anchor/Anchor.razor.cs
components/carousel/Carousel.razor.cs
components/core/Component/Overlay/Overlay.razor.cs
components/core/Component/Overlay/OverlayTrigger.razor.cs
components/core/JsInterop/DomEventService.cs
components/descriptions/Descriptions.razor.cs
components/dropdown/DropdownButton.cs
components/grid/Row.razor.cs
components/input/Input.cs
components/input/TextArea.razor.cs
components/layout/Sider.razor.cs
components/list/ListItem.razor.cs
components/select/Select.razor.cs
components/select/internal/SelectContent.razor.cs
components/slider/Slider.razor.cs
components/table/Table.razor.cs
components/tabs/Tabs.razor.cs

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] Changelog is provided or not needed

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/ant-design-blazor/ant-design-blazor?pullRequest=1857) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/ant-design-blazor/ant-design-blazor?pullRequest=1857) before we can accept your contribution.<br/><br/>:x: tonyyip1969 [sign now](https://cla.dotnetfoundation.org/ant-design-blazor/ant-design-blazor?pullRequest=1857)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/ant-design-blazor/ant-design-blazor?pullRequest=1857) it.</sub>

**Comment by ElderJames** [comment]
Thanks for contribute this @tonyyip1969 , please sign the CLA first.

**Comment by ElderJames** [comment]
So we're directly reusing subscriptions, right?

**Comment by tonyyip1969** [comment]
You are right. This may be the possible solution without much affected the component subscribed javascript callback.

**Comment by tonyyip1969** [comment]
The below code extracted from OverlayTrigger.razor.cs, the user may confuse which way/method should be used for register callback
1. DomEventService.AddEventListener("document", "mouseup", OnMouseUp);
or
2. _domEventListener.Add(Ref, "click", OnUnboundClick);

should I change the method name to
1. DomEventService.AddShareEventListener("document", "mouseup", OnMouseUp);
or
2. _domEventListener.AddExclusiveEventListerner(Ref, "click", OnUnboundClick);

any suggestion?

`
protected override void OnAfterRender(bool firstRender)
{
    if (firstRender)
    {
        DomEventService.AddEventListener("document", "mouseup", OnMouseUp);
        DomEventService.AddEventListener("window", "resize", OnWindowResize);
        DomEventService.AddEventListener("document", "scroll", OnWindowScroll);
    }

    base.OnAfterRender(firstRender);
}

protected override Task OnAfterRenderAsync(bool firstRender)
{
	if (firstRender && Unbound != null)
	{
		Ref = RefBack.Current;
		
		_domEventListener.Add(Ref, "click", OnUnboundClick);
		_domEventListener.Add(Ref, "mouseover", OnUnboundMouseEnter);
		_domEventListener.Add(Ref, "mouseout", OnUnboundMouseLeave);
		_domEventListener.Add(Ref, "focusin", OnUnboundFocusIn);
		_domEventListener.Add(Ref, "focusout", OnUnboundFocusOut);
		_domEventListener.Add(Ref, "contextmenu", OnContextMenu, true);
	}
	return base.OnAfterRenderAsync(firstRender);
}`

File: components/core/JsInterop/DomEventListener.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        private readonly IJSRuntime _jsRuntime;
+        private Dictionary<string, DotNetObjectReference<Invoker<T>>> _dotNetObjectStore = new();
+
+        public DomEventListener(IJSRuntime jsRuntime)
+        {
+            _jsRuntime = jsRuntime;
+        }
+
+        public void Add(object dom, string eventName, Action<T> callback, bool preventDefault = false)
```
Comment (ElderJames) [inline components/core/JsInterop/DomEventListener.cs:18]: Can we use the generic type at `Add` method instead of class name?

Comment (ElderJames) [reply components/core/JsInterop/DomEventListener.cs:18]: Because we may use more than one return type at a time.

Comment (tonyyip1969) [reply components/core/JsInterop/DomEventListener.cs:18]: Look like not possible, VS doesn't know what is the generic type for _dotNetObjectStore.

image

Comment (ElderJames) [reply components/core/JsInterop/DomEventListener.cs:18]: Maybe can use `Dictionary<string, IDisposable>`.

BTW, I think we also need to add the Exclusive logic.

Comment (tonyyip1969) [reply components/core/JsInterop/DomEventListener.cs]: `Dictionary<string, IDisposable>` works perfectly as expected. Pushed amended code and now with Exclusive and reuse logic in the same class/service.

Comment (ElderJames) [reply components/core/JsInterop/DomEventListener.cs]: Thank you!

Comment (tonyyip1969) [reply components/core/JsInterop/DomEventListener.cs]: I noticed there were failures on the bunit test. Do I need to do something about it?

File: components/core/JsInterop/DomEventListener.cs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                        subscription.Delegate.DynamicInvoke(tP);
+                    }
+                }));
+
+                _jsRuntime.InvokeAsync<string>(JSInteropConstants.AddDomEventListener, dom, eventName, preventDefault, dotNetObject);
+            }
+            _domEventSubscriptionsStore[key].Add(new DomEventSubscription(callback, typeof(T)));
+        }
+
+        public void RemoveShared<T>(object dom, string eventName, Action<T> callback)
```
Comment (ElderJames) [inline components/core/JsInterop/DomEventListener.cs:87]: Can the RemoveShared method be used without a generic type? That would be more convenient.

Comment (ElderJames) [reply components/core/JsInterop/DomEventListener.cs:87]: Change `Action<T>` to Delegate

Comment (tonyyip1969) [reply components/core/JsInterop/DomEventListener.cs:87]: image

image

I have encountered the error above after changed to Delegate. Any suggestion? I do not quite understand how the Delegate work.

Comment (tonyyip1969) [reply components/core/JsInterop/DomEventListener.cs:88]: @ElderJames , added DisposeShared method to remove all the callback in a single statement

Comment (ElderJames) [reply components/core/JsInterop/DomEventListener.cs:88]: Cool!

**Review by ElderJames — APPROVED** [review]
LGTM. This helps a lot, thank you!

**Comment by tonyyip1969** [comment]
Hi @ElderJames, thanks for approved my PR. What should I do on the 1 failing check?

**Comment by ElderJames** [comment]
@tonyyip1969 It's related the test project, I'll fix it.

**Comment by ElderJames** [comment]
Hi @tonyyip1969 , we found many issue at the DomEventListener disponse. I think it was because the `Dispose()` did not call RemoveEventListener and it continues to trigger operations on the removed DOM.

issues: #2393 #2136

**Pull request merged**